File: CSharp/DataStructures/Queue.cs
using System;
using System.Collections;
using System.Collections.Generic;

public class Queue<T> : IEnumerable<T>
{
    public class Node
    {
        public T value;
        public Node next;

        public Node(T value, Node next)
        {
            this.value = value;
            this.next = next;
        }
    }

    private Node head;
    private Node tail;
    public int Size { get; private set; }

    public bool IsEmpty()
    {
        return Size <= 0;
    }

    public T Peek()
    {
        if (IsEmpty())
        {
            throw new InvalidOperationException();
        }
        return head.value;
    }

    public void Enqueue(T value)
    {
        if (IsEmpty())
        {
            tail = new Node(value, null);
            head = tail;
            Size++;
            return;
        }
        tail.next = new Node(value, null);
        tail = tail.next;
        Size++;
        return;
    }

    public T Dequeue()
    {
        if (IsEmpty())
        {
            throw new InvalidOperationException();
        }
        if(Size == 1)
        {
            Node tmp = head;
            head = null;
            tail = null;
            Size = 0;
            return tmp.value;
        }
        else
        {
            Node tmp = head;
            head = head.next;
            Size--;
            return tmp.value;
        }
    }

    public T Poll()
    {
        if (IsEmpty())
        {
            return default(T);
        }
        if(Size == 1)
        {
            Node tmp = head;
            head = null;
            tail = null;
            Size = 0;
            return tmp.value;
        }
        else
        {
            Node tmp = head;
            head = head.next;
            Size--;
            return tmp.value;
        }
    }

    override
    public string ToString()
    {
        string res = "{";
        Node tmp = head;
        while(tmp != null)
        {
            res += tmp.value.ToString();
            if (tmp.next != null) res += ", ";
            tmp = tmp.next;
        }
        res += "}";
        return res;
    }

    public IEnumerator<T> GetEnumerator()
    {
        Node node = this.head;
        while (node != null)
        {
            yield return node.value;
            node = node.next;
        }
    }

    IEnumerator IEnumerable.GetEnumerator()
    {
        return this.GetEnumerator();
    }
}

File: PHP/DataStructures/Queue.php
<?php

class Queue {

    // Head, tail, and size of queue.
    private $head;
    private $tail;
    private $size;

    public function isEmpty() {
        return  $this->size <=0;
    }
    
    public function peek() {
        if($this->isEmpty()) {
            throw new Exception("Queue is empty!");
        }
        return $this->head->value;
    }

    public function enqueue($value) {
        if($this->isEmpty()) {
            $this->tail = new Node($value, null);
            $this->head = $this->tail;
            $this->size++;
        }
        else {
            $this->tail->next = new Node($value, null);
            $this->tail = $this->tail->next;
            $this->size++;
        }
    }

    public function dequeue() {
        if($this->isEmpty()) {
            throw new Exception("Queue is empty!");
        }
        if($this->size == 1) {
            $tmp = $this->head;
            $this->head = null;
            $this->tail = null;
            $this->size = 0;
            return $tmp->value;
        }
        else {
            $tmp = $this->head;
            $this->head = $this->head->next;
            $this->size--;
            return $tmp->value;
        }
    }

    public function poll() {
        if($this->isEmpty()) {
            return null;
        }
        if($this->size == 1) {
            $tmp = $this->head;
            $this->head = null;
            $this->tail = null;
            $this->size = 0;
            return $tmp->value;
        }
        else {
            $tmp = $this->head;
            $this->head = $this->head->next;
            $this->size--;
            return $tmp->value;
        }
    }

    public function size() {
        return $this->size;
    }

    public function __toString() {
        $res = "{";
        $tmp = $this->head;
        while($tmp) {
            $res .= (string) $tmp->value;
            if($tmp->next) {
                $res .= ", ";
            }
            $tmp = $tmp->next;
        }
        return $res;
    }
}

// Supporting Node class.
class Node {
    public $value;
    public $next;

    public function __construct($value, $next) {
        $this->value = $value;
        $this->next = $next;
    }
}
